tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java

```diff
@@ -126,47 +118,49 @@ protected void channelRead0(final ChannelHandlerContext ctx, final Request reque
         if (handler == null) {
             handler = topChannel.getDefaultUserHandler();
         }
-
         if (handler == null) {
-            sendError(ErrorType.BadRequest,
-                "No handler function for service:endpoint=" + service + ":" + endpoint,
-                request, ctx);
+            sendError(
+                ErrorType.BadRequest,
+                "No handler function for service:endpoint=" + service + ':' + endpoint,
+                request,
+                ctx
+            );
             return;
         }
 
-        ListenableFuture<Response> responseFuture;
-        // In case of an AsyncRequestHandler no need to submit a task on the executor. It does
-        // require a down-cast to AsyncRequestHandler.
-        if (handler instanceof AsyncRequestHandler) {
-            AsyncRequestHandler asyncHandler = (AsyncRequestHandler) handler;
-            responseFuture = sendRequestToAsyncHandler(asyncHandler, request);
-        } else {
-            // Handle the request in a separate thread and get a future to it
-            responseFuture = listeningExecutorService.submit(
-                new CallableHandler(handler, topChannel, request));
+        ListenableFuture<? extends Response> responseFuture;
+        try {
+            // In case of an AsyncRequestHandler there's no need to submit a task on the executor.
+            // It does require a down-cast to AsyncRequestHandler.
```

Comment:
This whole block seems weird design (I realise you only mostly reformatted and added the try block

Reply:
that's an artifact of dealing with essentially two separate interfaces: original `RequestHandler` and `AsyncRequestHandler` that was added later.